omniledger/service/viewchange.go

```diff
@@ -0,0 +1,371 @@
+// function should only be used as a callback in ViewChangeLog.
+func (s *Service) sendViewChangeReq(view viewchange.View) error {
+	log.Lvl2(s.ServerIdentity(), "sending view-change request for view:", view)
+	latest, err := s.db().GetLatestByID(view.ID)
+	if err != nil {
+		return err
+	}
+	req := viewchange.InitReq{
+		SignerID: s.ServerIdentity().ID,
+		View:     view,
+	}
+	if err := req.Sign(s.getPrivateKey()); err != nil {
+		return err
+	}
+	for _, sid := range latest.Roster.List {
+		if sid.Equal(s.ServerIdentity()) {
+			continue
+		}
+		if err := s.SendRaw(sid, &req); err != nil {
+			// Having an error here is fine because not all the
+			// nodes are guaranteed to be online. So we log a
+			// warning instead of returning an error.
+			log.Warn(s.ServerIdentity(), err)
+		}
+	}
+	return nil
+}
+
+func (s *Service) sendNewView(proof []viewchange.InitReq) {
+
+	if len(proof) == 0 {
+		log.Error(s.ServerIdentity(), "not enough proofs")
+	}
+	log.Lvl2(s.ServerIdentity(), "sending new-view request for view:", proof[0].View)
+
+	// Our own proof might not be signed, so sign it.
+	for i := range proof {
+		if proof[i].SignerID.Equal(s.ServerIdentity().ID) && len(proof[i].Signature) == 0 {
+			proof[i].Sign(s.getPrivateKey())
+		}
+	}
+
+	sb := s.db().GetByID(proof[0].View.ID)
+	req := viewchange.NewViewReq{
+		Roster: *rotateRoster(sb.Roster, proof[0].View.LeaderIndex),
+		Proof:  proof,
+	}
+
+	go func() {
+		// This go-routine eventually exists because both cosi and
+		// block creation have a timeout.
+		sig, err := s.startViewChangeCosi(req)
+		if err != nil {
+			log.Error(s.ServerIdentity(), err)
+			return
+		}
+		if len(sig) == 0 {
+			log.Error(s.ServerIdentity(), "empty viewchange cosi signature")
+			return
+		}
+		if err := s.createViewChangeBlock(req, sig); err != nil {
+			log.Error(s.ServerIdentity(), err)
+		}
+	}()
+}
+
+func (s *Service) computeInitialDuration(scID skipchain.SkipBlockID) time.Duration {
+	interval, err := s.LoadBlockInterval(scID)
+	if err != nil {
+		panic("we should only computer the interval on a known block," +
+			" otherwise it is a progarmmer error: " + err.Error())
+	}
+	return rotationWindow * interval
+}
+
+func (s *Service) getFaultThreshold(sbID skipchain.SkipBlockID) int {
+	sb := s.db().GetByID(sbID)
+	return len(sb.Roster.List) / 3
+}
+
+// handleViewChangeReq should be registered as a handler for viewchange.InitReq
+// messages.
+func (s *Service) handleViewChangeReq(env *network.Envelope) {
+	// Parse message.
+	req, ok := env.Msg.(*viewchange.InitReq)
+	if !ok {
+		log.Error(s.ServerIdentity(), "failed to cast to viewchange.ViewChangeReq")
+		return
+	}
+	// Should not be sending to ourselves.
+	if req.SignerID.Equal(s.ServerIdentity().ID) {
+		log.Error(s.ServerIdentity(), "should not send to ourselve")
+		return
+	}
+
+	// Check that the genesis exists and the view is valid.
+	_, err := s.db().GetLatestByID(req.View.Gen)
```

Comment:
This is a place where untrusted input can come from the network, right? Let's be really careful.

Why are we using GetLatestByID here, and throwing away the result? On a long chain, it is expensive, because it traverses the entire chain. Why not GetByID and then check that block.Index == 0?

Reply:
Actually, there's a cache in skipchain, and `GetLatest` should usually loads from the cache. But you're right that `GetByID` is enough here.